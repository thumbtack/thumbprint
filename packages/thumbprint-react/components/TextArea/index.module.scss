@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

// The `.root.root.root` is used to increase specificity so Thumbprint takes precedence over a
// highly specific selector in Thumbtack's website.
// https://github.com/thumbtack/thumbprint-archive/issues/244
.root.root.root {
    font-family: $tp-font-family__base;
    font-size: $tp-font__body__1__size;
    line-height: $tp-font__body__1__line-height;
    width: 100%;
    border-width: 1px;
    border-style: solid;
    border-radius: $tp-border-radius__base;
    padding: 10px 14px;
    vertical-align: middle;
    resize: vertical;
    height: 152px;
    min-height: 104px;

    &:focus {
        color: $tp-color__black;
        border-color: $tp-color__blue;
    }

    ::-webkit-input-placeholder,
    ::-moz-placeholder,
    :-ms-input-placeholder {
        color: $tp-color__black-300;
    }

    &StateDefault {
        border-color: $tp-color__gray;
        color: $tp-color__black;
        background-color: $tp-color__white;
    }

    &StateDisabled {
        border-color: $tp-color__gray;
        color: $tp-color__gray;
        background-color: $tp-color__gray-200;
    }

    &StateError {
        border-color: $tp-color__red;
        color: $tp-color__red;
        background-color: $tp-color__white;

        // This must be three separate rules, not a comma-separated list, to avoid the different
        // vendor prefixes conflicting.
        &::-webkit-input-placeholder {
            color: $tp-color__red;
        }

        &::-moz-placeholder {
            color: $tp-color__red;
        }

        &:-ms-input-placeholder {
            color: $tp-color__red;
        }
    }
}
