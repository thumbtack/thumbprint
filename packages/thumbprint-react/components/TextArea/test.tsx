import React from 'react';
import { shallow, mount, render } from 'enzyme';
import noop from 'lodash/noop';
import TextArea from './index';

test('adds placeholder', () => {
    const wrapper = mount(<TextArea placeholder="Goose" onChange={jest.fn} value="goose" />);
    expect(wrapper.prop('placeholder')).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('renders an error state', () => {
    const wrapper = mount(<TextArea hasError onChange={jest.fn} value="goose" />);
    expect(wrapper).toMatchSnapshot();
});

test('adds `id` attribute', () => {
    const wrapper = mount(<TextArea id="Goose" onChange={jest.fn} value="goose" />);
    expect(wrapper.find('textarea').prop('id')).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('adds `disabled` attribute', () => {
    const wrapper = mount(<TextArea isDisabled onChange={jest.fn} value="goose" />);
    expect(wrapper.find('textarea').prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `readOnly` attribute', () => {
    const wrapper = mount(<TextArea isReadOnly onChange={jest.fn} value="goose" />);
    expect(wrapper.find('textarea').prop('readOnly')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `required` attribute', () => {
    const wrapper = mount(<TextArea isRequired onChange={jest.fn} value="goose" />);
    expect(wrapper.find('textarea').prop('required')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `maxLength` attribute', () => {
    const wrapper = mount(<TextArea maxLength={100} onChange={jest.fn} value="goose" />);
    expect(wrapper.prop('maxLength')).toBe(100);
    expect(wrapper).toMatchSnapshot();
});

test('adds `name` HTML attribute', () => {
    const wrapperA = mount(<TextArea name="duck" onChange={noop} value="goose" />);
    expect(wrapperA.find('textarea').prop('name')).toEqual('duck');
    expect(wrapperA).toMatchSnapshot();
});

test('renders `value` as the `value` attribute', () => {
    const wrapper = render(<TextArea value="Goose" onChange={jest.fn} />);
    expect(wrapper.text()).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('calls `onChange` function and passes value', () => {
    const onChange = jest.fn();

    const wrapper = mount(<TextArea onChange={onChange} value="goose" />);

    const firstEvent = { target: { value: 'He' } };
    const secondEvent = { target: { value: 'Hello' } };

    wrapper.find('textarea').simulate('change', firstEvent);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenLastCalledWith(
        firstEvent.target.value,
        expect.objectContaining({
            target: expect.any(Object),
        }),
    );

    wrapper.find('textarea').simulate('change', secondEvent);
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenLastCalledWith(
        secondEvent.target.value,
        expect.objectContaining({
            target: expect.any(Object),
        }),
    );
});

test('calls `onBlur` function when supplied', () => {
    const onBlur = jest.fn();

    const wrapper = shallow(<TextArea onBlur={onBlur} onChange={jest.fn} value="goose" />);

    wrapper.simulate('blur');
    expect(onBlur).toHaveBeenCalledTimes(1);
});

test('calls `onFocus` function when supplied', () => {
    const onFocus = jest.fn();

    const wrapper = shallow(<TextArea onFocus={onFocus} onChange={jest.fn} value="goose" />);

    wrapper.simulate('focus');
    expect(onFocus).toHaveBeenCalledTimes(1);
});

test('adds `dataTest` prop', () => {
    const wrapper = mount(<TextArea dataTest="Duck" onChange={noop} value="goose" />);
    expect(wrapper.find('textarea').prop('data-test')).toEqual('Duck');
    expect(wrapper).toMatchSnapshot();
});
