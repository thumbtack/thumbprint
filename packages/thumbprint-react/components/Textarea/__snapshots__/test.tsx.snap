// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<Textarea
  dataTest="Duck"
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateDefault"
    data-test="Duck"
    disabled={false}
    onChange={[Function]}
    required={false}
    value="goose"
  />
</Textarea>
`;

exports[`adds \`disabled\` attribute 1`] = `
<Textarea
  isDisabled={true}
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateDisabled"
    disabled={true}
    onChange={[Function]}
    required={false}
    value="goose"
  />
</Textarea>
`;

exports[`adds \`id\` attribute 1`] = `
<Textarea
  id="Goose"
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateDefault"
    disabled={false}
    id="Goose"
    onChange={[Function]}
    required={false}
    value="goose"
  />
</Textarea>
`;

exports[`adds \`maxLength\` attribute 1`] = `
<Textarea
  maxLength={100}
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateDefault"
    disabled={false}
    maxLength={100}
    onChange={[Function]}
    required={false}
    value="goose"
  />
</Textarea>
`;

exports[`adds \`name\` HTML attribute 1`] = `
<Textarea
  name="duck"
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateDefault"
    disabled={false}
    name="duck"
    onChange={[Function]}
    required={false}
    value="goose"
  />
</Textarea>
`;

exports[`adds \`required\` attribute 1`] = `
<Textarea
  isRequired={true}
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateDefault"
    disabled={false}
    onChange={[Function]}
    required={true}
    value="goose"
  />
</Textarea>
`;

exports[`adds placeholder 1`] = `
<Textarea
  onChange={[Function]}
  placeholder="Goose"
  value="goose"
>
  <textarea
    className="root rootStateDefault"
    disabled={false}
    onChange={[Function]}
    placeholder="Goose"
    required={false}
    value="goose"
  />
</Textarea>
`;

exports[`renders \`value\` as the \`value\` attribute 1`] = `
<textarea
  class="root rootStateDefault"
>
  Goose
</textarea>
`;

exports[`renders an error state 1`] = `
<Textarea
  hasError={true}
  onChange={[Function]}
  value="goose"
>
  <textarea
    className="root rootStateError"
    disabled={false}
    onChange={[Function]}
    required={false}
    value="goose"
  />
</Textarea>
`;
